File: app/lib/core/mentions.dart
import 'package:campfire/models/channel.dart';
import 'package:campfire/models/user.dart';
import 'package:meta/meta.dart';

/// Port of `lib/mentions.ts`, character for character — the two clients have to
/// linkify the same text the same way, or the same message reads differently
/// depending on where it is opened.
const everyoneMention = 'everyone';

/// Matches the same charset/length auth enforces for usernames (see
/// `server/app/schemas/auth.py`), plus the "everyone" pseudo-mention.
///
/// Channel names have no such restriction (free text, see
/// `server/app/models/channel.py`), so `#` just grabs the next run of
/// non-whitespace and it is checked against known channel names below — this
/// only linkifies cleanly for the common space-free "kebab-case" convention.
///
/// Both require a word boundary before the trigger, so `foo@bar` or a literal
/// `#` in prose is not misread as a mention.
final _mentionPattern = RegExp(
  r'(?<![\w@])@(everyone|[a-zA-Z0-9_]{3,32})\b'
  '|'
  r'(?<![\w#])#(\S+)',
);

/// Trailing punctuation that should not be swallowed into a channel name.
final _trailingPunctuation = RegExp(r'[.,!?;:)]+$');

enum MentionKind { everyone, user, channel }

/// A run of message text: plain when [mention] is null, otherwise the whole
/// `@name` or `#channel` token including its trigger character.
@immutable
class MentionSegment {
  const MentionSegment(this.text, {this.mention, this.channelId});

  final String text;
  final MentionKind? mention;

  /// Set for [MentionKind.channel] — what to open when the segment is tapped.
  final String? channelId;

  @override
  String toString() => 'MentionSegment($text, $mention)';

  @override
  bool operator ==(Object other) =>
      other is MentionSegment &&
      other.text == text &&
      other.mention == mention &&
      other.channelId == channelId;

  @override
  int get hashCode => Object.hash(text, mention, channelId);
}

/// Splits message content into plain-text and mention segments for rendering.
///
/// An `@name` that belongs to nobody, or a `#thing` that is not a channel,
/// stays plain text — the same call the web client makes, so an unknown handle
/// does not get highlighted as if it were real.
List<MentionSegment> splitMentions(
  String content,
  Set<String> knownUsernames,
  Map<String, Channel> channelsByName,
) {
  final segments = <MentionSegment>[];
  var lastIndex = 0;

  for (final match in _mentionPattern.allMatches(content)) {
    final userOrEveryone = match.group(1);
    final channelToken = match.group(2);

    MentionSegment? segment;
    var matchedLength = match.group(0)!.length;

    if (userOrEveryone != null) {
      final isKnown = userOrEveryone == everyoneMention ||
          knownUsernames.contains(userOrEveryone.toLowerCase());
      if (isKnown) {
        segment = MentionSegment(
          match.group(0)!,
          mention: userOrEveryone == everyoneMention
              ? MentionKind.everyone
              : MentionKind.user,
        );
      }
    } else if (channelToken != null) {
      final trimmed = channelToken.replaceFirst(_trailingPunctuation, '');
      final channel = channelsByName[trimmed.toLowerCase()];
      if (channel != null) {
        matchedLength = 1 + trimmed.length;
        segment = MentionSegment(
          '#$trimmed',
          mention: MentionKind.channel,
          channelId: channel.id,
        );
      }
    }

    if (segment == null) continue;

    if (match.start > lastIndex) {
      segments.add(MentionSegment(content.substring(lastIndex, match.start)));
    }
    segments.add(segment);
    lastIndex = match.start + matchedLength;
  }

  if (lastIndex < content.length) {
    segments.add(MentionSegment(content.substring(lastIndex)));
  }
  return segments;
}

/// Whether [content] mentions [username] directly or via `@everyone`. Drives
/// the highlight on the message row and, later, the notification.
bool messageMentionsUser(String content, String username) {
  final lower = username.toLowerCase();
  for (final match in _mentionPattern.allMatches(content)) {
    final name = match.group(1);
    if (name != null && (name == everyoneMention || name.toLowerCase() == lower)) {
      return true;
    }
  }
  return false;
}

enum MentionTrigger {
  user('@'),
  channel('#');

  const MentionTrigger(this.character);

  final String character;
}

/// An `@query` or `#query` being typed right now.
@immutable
class MentionQuery {
  const MentionQuery({required this.trigger, required this.query, required this.start});

  final MentionTrigger trigger;

  /// Text typed after the trigger, used to filter candidates.
  final String query;

  /// Index in the field's value where the trigger character starts.
  final int start;

  @override
  bool operator ==(Object other) =>
      other is MentionQuery &&
      other.trigger == trigger &&
      other.query == query &&
      other.start == start;

  @override
  int get hashCode => Object.hash(trigger, query, start);
}

/// Detects an in-progress mention immediately before the cursor, if any.
MentionQuery? activeMentionQuery(String value, int cursor) {
  final upToCursor = value.substring(0, cursor.clamp(0, value.length));

  var start = -1;
  MentionTrigger? trigger;
  for (var i = upToCursor.length - 1; i >= 0; i--) {
    final ch = upToCursor[i];
    if (ch == '@' || ch == '#') {
      start = i;
      trigger = ch == '@' ? MentionTrigger.user : MentionTrigger.channel;
      break;
    }
    if (RegExp(r'\s').hasMatch(ch)) break;
  }
  if (start == -1 || trigger == null) return null;

  // Only at a word boundary — not mid-token, e.g. "foo@bar".
  if (start > 0) {
    final before = upToCursor[start - 1];
    if (RegExp(r'\w').hasMatch(before) || before == trigger.character) return null;
  }

  return MentionQuery(
    trigger: trigger,
    query: upToCursor.substring(start + 1),
    start: start,
  );
}

/// One row of the autocomplete list.
class MentionCandidate {
  const MentionCandidate({required this.key, required this.label, required this.insert});

  /// Null for `@everyone`, which is not a person — the list uses it to decide
  /// between an avatar and a glyph.
  final String key;
  final String label;
  final String insert;
}

/// Up to eight matches for what is being typed, prefix-matched and in the order
/// the web client offers them: `everyone` first, then users, or text channels
/// for `#`.
List<MentionCandidate> mentionCandidates(
  MentionTrigger trigger,
  String query,
  List<User> users,
  List<Channel> channels,
) {
  final lower = query.toLowerCase();

  if (trigger == MentionTrigger.channel) {
    return channels
        .where((c) => c.type == ChannelType.text && c.name.toLowerCase().startsWith(lower))
        .map((c) => MentionCandidate(key: c.id, label: c.name, insert: c.name))
        .take(8)
        .toList();
  }

  return [
    if (everyoneMention.startsWith(lower))
      const MentionCandidate(
        key: everyoneMention,
        label: everyoneMention,
        insert: everyoneMention,
      ),
    for (final user in users)
      if (user.username.toLowerCase().startsWith(lower))
        MentionCandidate(key: user.id, label: user.username, insert: user.username),
  ].take(8).toList();
}

File: client/src/lib/mentions.ts
import type { Channel, User } from "@/lib/types";

export const EVERYONE_MENTION = "everyone";

// Matches the same charset/length auth enforces for usernames (see
// server/app/schemas/auth.py), plus the "everyone" pseudo-mention.
// Channel names have no such charset restriction (free text, see
// server/app/models/channel.py), so "#" just grabs the next run of
// non-whitespace and we check it against known channel names below —
// this only linkifies cleanly for the common space-free "kebab-case"
// channel naming convention.
// Both require a word boundary before the trigger, so "foo@bar" or a
// literal "#" in prose doesn't get misread as a mention.
const MENTION_PATTERN =
  "(?<![\\w@])@(everyone|[a-zA-Z0-9_]{3,32})\\b" + "|" + "(?<![\\w#])#(\\S+)";

function mentionsRegex(): RegExp {
  return new RegExp(MENTION_PATTERN, "g");
}

export interface MentionSegment {
  text: string;
  mention: "everyone" | "user" | "channel" | null;
  channelId?: string;
}

/** Splits message content into plain-text and mention segments for rendering. */
export function splitMentions(
  content: string,
  knownUsernames: Set<string>,
  channelsByName: Map<string, Channel>,
): MentionSegment[] {
  const segments: MentionSegment[] = [];
  const regex = mentionsRegex();
  let lastIndex = 0;
  let match: RegExpExecArray | null;

  while ((match = regex.exec(content)) !== null) {
    const [full, userOrEveryone, channelToken] = match;
    let segment: MentionSegment | null = null;
    let matchedLength = full.length;

    if (userOrEveryone !== undefined) {
      const isKnown = userOrEveryone === EVERYONE_MENTION || knownUsernames.has(userOrEveryone.toLowerCase());
      if (isKnown) {
        segment = { text: full, mention: userOrEveryone === EVERYONE_MENTION ? "everyone" : "user" };
      }
    } else if (channelToken !== undefined) {
      // Channel names are free text, so trailing punctuation like "#chat," or
      // "#chat." shouldn't be swallowed into the match — trim it off first.
      const trimmedToken = channelToken.replace(/[.,!?;:)]+$/, "");
      const channel = channelsByName.get(trimmedToken.toLowerCase());
      if (channel) {
        matchedLength = 1 + trimmedToken.length;
        segment = { text: `#${trimmedToken}`, mention: "channel", channelId: channel.id };
      }
    }

    if (!segment) continue;

    if (match.index > lastIndex) {
      segments.push({ text: content.slice(lastIndex, match.index), mention: null });
    }
    segments.push(segment);
    lastIndex = match.index + matchedLength;
  }

  if (lastIndex < content.length) {
    segments.push({ text: content.slice(lastIndex), mention: null });
  }
  return segments;
}

/** Whether `content` mentions `username` directly or via @everyone. */
export function messageMentionsUser(content: string, username: string): boolean {
  const lowerUsername = username.toLowerCase();
  const regex = mentionsRegex();
  let match: RegExpExecArray | null;
  while ((match = regex.exec(content)) !== null) {
    const name = match[1];
    if (name !== undefined && (name === EVERYONE_MENTION || name.toLowerCase() === lowerUsername)) {
      return true;
    }
  }
  return false;
}

export type MentionTrigger = "@" | "#";

export interface MentionQuery {
  trigger: MentionTrigger;
  /** Text typed after the trigger, used to filter candidates. */
  query: string;
  /** Index in the textarea value where the trigger character starts. */
  start: number;
}

/** Detects an in-progress "@query" or "#query" token immediately before the cursor, if any. */
export function activeMentionQuery(value: string, cursor: number): MentionQuery | null {
  const upToCursor = value.slice(0, cursor);

  let start = -1;
  let trigger: MentionTrigger | null = null;
  for (let i = upToCursor.length - 1; i >= 0; i--) {
    const ch = upToCursor[i];
    if (ch === "@" || ch === "#") {
      start = i;
      trigger = ch;
      break;
    }
    if (/\s/.test(ch)) break;
  }
  if (start === -1 || !trigger) return null;

  // Only trigger at a word boundary — not mid-token, e.g. "foo@bar".
  const charBefore = upToCursor[start - 1];
  if (charBefore !== undefined && (/\w/.test(charBefore) || charBefore === trigger)) return null;

  return { trigger, query: upToCursor.slice(start + 1), start };
}

export interface MentionCandidate {
  key: string;
  label: string;
  insert: string;
}

export function mentionCandidates(
  trigger: MentionTrigger,
  query: string,
  users: User[],
  channels: Channel[],
  excludedUserId?: string,
): MentionCandidate[] {
  const lowerQuery = query.toLowerCase();

  if (trigger === "#") {
    return channels
      .filter((c) => c.type === "text" && c.name.toLowerCase().startsWith(lowerQuery))
      .map((c) => ({ key: c.id, label: c.name, insert: c.name }))
      .slice(0, 8);
  }

  const candidates: MentionCandidate[] = [];
  if (EVERYONE_MENTION.startsWith(lowerQuery)) {
    candidates.push({ key: "everyone", label: "everyone", insert: "everyone" });
  }
  for (const user of users) {
    if (user.id !== excludedUserId && user.username.toLowerCase().startsWith(lowerQuery)) {
      candidates.push({ key: user.id, label: user.username, insert: user.username });
    }
  }
  return candidates.slice(0, 8);
}
